### ISSUE TITLE
Increasing CPU usage on mouse movement

### ISSUE THREAD

**Issue by nvlled** [TolikPylypchuk/SharpHook#112]
On linux, the CPU usage keeps gradually increasing when I move my mouse, until it hits around 100%.

For instance, when I stop moving my mouse at 50% usage, it goes back to zero. Then when I move my mouse again, it resumes from 50% and continues on.

Not sure yet if this is on the C# part, or on libuiohook.

See this project https://github.com/nvlled/sharphook-bug to reproduce the bug.

There's also a video https://github.com/nvlled/sharphook-bug/blob/main/bug-demo.mp4 showing the bug.

**Comment by TolikPylypchuk** [comment]
Hello! Thanks for posting this issue! Well, this is certainly interesting. I will look into it in the nearest future.

**Comment by nvlled** [comment]
I built libuiohook from scratch and it seems the problem occurs as well with `demo_hook.c`.

I also commented out the call to `dispatch_mouse_release` in https://github.com/TolikPylypchuk/libuiohook/blob/c7500ce3b96756cddfd486e819ae576d6b05f1f5/src/x11/input_hook.c#L122, and the problem still occurs. Which means the problem isn't specific to mouse movement.

**Comment by nvlled** [comment]
Closing the issue because I found out that this bug is not specific to either SharpHook or libuiohook, but on the x11 extensions library.

I copied the minimum relevant code from input_hook.c to reproduce the bug: https://github.com/nvlled/xrecord-enable-context Although, I'm still trying to see if this is just something specific on my system.

**Comment by TolikPylypchuk** [comment]
OK, no problem, if it's an issue with libuiohook after all, feel free to reopen this issue.

**Comment by nvlled** [comment]
After poking around libxtst, I found the reason. https://gitlab.freedesktop.org/xorg/lib/libxtst/-/blob/master/src/XRecord.c?ref_type=heads#L540
```c
    for (rbp = cache->reply_buffers; rbp; rbp = rbp->next) {
	if (rbp->ref_count == 0) {
	    if (rbp->nbytes >= nbytes)
		return rbp;
	    else
		saved_rb = rbp;
	}
    }
```
This loop gets larger and larger as I move my mouse. At first, I thought this was a bug in the libxtst and they forgot to call XRecordFreeData somewhere. But searching other projects that uses libxtst, it seems the caller is responsible for freeing the allocated record_data.

Looking up the docs confirms this: https://www.x.org/releases/X11R7.7/doc/libXtst/recordlib.html

> Recording clients should use the XRecordFreeDatafunction to free the XRecordInterceptData structure.

I opened up PR that fixes the bug https://github.com/TolikPylypchuk/libuiohook/pull/4

**Comment by nvlled** [comment]
@TolikPylypchuk not sure how to re-open this issue though

**Comment by TolikPylypchuk** [comment]
Thanks for looking into it! I've merged the pull request and will release a new version of SharpHook in the coming days (probably tomorrow).

**Comment by TolikPylypchuk** [comment]
I've just released version 5.3.7 which includes your pull request to libuiohook, so this issue can be closed.
